python_ta/checkers/bad_loop_indexing_checker.py

```diff
@@ -0,0 +1,109 @@
+"""checker for bad indexing in a loop.
+"""
+
+import astroid
+from pylint.interfaces import IAstroidChecker
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import check_messages
+from typing import List, Optional
+from astroid.node_classes import NodeNG
+
+
+class BadLoopIndexingChecker(BaseChecker):
+    __implements__ = IAstroidChecker
+    # name is the same as file name but without _checker part
+    name = 'bad_loop_indexing'
+    # use dashes for connecting words in message symbol
+    msgs = {'E9994': (f'Bad use of loop variable %s in for loop',
+                      'bad-loop-indexing',
+                      'Used when you have an loop variable in a for loop '
+                      'where its only usage is to index the iterable'),
+            }
+
+    # this is important so that your checker is executed before others
+    priority = -1
+
+    # pass in message symbol as a parameter of check_messages
+    @check_messages("bad-loop-indexing")
+    def visit_for(self, node: astroid.For) -> None:
+        if self._is_bad_loop_indexing(node):
+            args = f"'{node.target.name}'"
+            self.add_message('bad-loop-indexing', node=node.target, args=args)
+
+    def _is_bad_loop_indexing(self, node: astroid.For) -> bool:
+        """Return whether the usage of the indexing variable in the for loop is ONLY to index the iterable.
+        True if bad, False if not bad.
+        """
+        # Store the iteration variable.
+        index = node.target.name
+
+        # Check if the iterable of the for loop is of the form "range(len(lst))".
+        iterable = _iterable_if_range(node.iter)
+        if iterable is None:
+            return False
+
+        # AugAssign is special (i += 2 is actually i = i + 2).
+        # Check if any AssignName nodes in an AugAssign node refers to the iteration variable.
+        if any(index_node.lookup(index)[1][0] == node.target for index_node in _augassign_index_nodes(node, index)):
+            return False
+        # i) For all index_node in augassign_index_nodes, .lookup != node.target, so they are all rebound
+        # or ii) there are no AugAssign nodes.
+        return not any(index_node.lookup(index)[1][0] == node.target
+                       for index_node in _good_use_index_nodes(node, index)) \
+            and (_index_assign_name_nodes(node, index) or _index_name_nodes(node, index))
+
+
+# Helper functions
+def _iterable_if_range(node: Optional[NodeNG]) -> Optional[str]:
+    """Return the string of the iterable if this node is of the form: range(len(iterable)).
+    Return None otherwise.
+    """
+    # Order of operations is important.
+    if isinstance(node, astroid.Call):
```

Comment:
There's no reason to nest the `if`s here. Just leave one with a long string of conditions. Also, use parentheses to enclose the different clauses, rather than using backslashes:

```
if (condition1 and
        condition2 and
        condition3 and
        ...
        conditionN):
```

Also, you're missing a final check for whether the inner `args[0]` is a `Name` node.